src/ra_utility

```diff
@@ -0,0 +1,241 @@
+
+template<typename Integral, class = std::enable_if_t<std::is_integral_v<Integral>>> _NODISCARD _CONSTANT_FN
+to_floating(_In_ Integral i) noexcept { return static_cast<float_type>(i); }
+
+template<typename FloatingPoint, class = std::enable_if_t<std::is_floating_point_v<FloatingPoint>>>
+_NODISCARD _CONSTANT_FN ftoi(_In_ FloatingPoint fp) noexcept
+{ 
+	if(std::signbit(fp))
+		return std::lround(fp); 
+	return to_unsigned(std::lround(fp));
+}
+
+template<typename Arithmetic, class = std::enable_if_t<std::is_arithmetic_v<Arithmetic>>> _NODISCARD _CONSTANT_FN
+sqr(_In_ Arithmetic a) noexcept { return std::pow(a, 2); }
+
+template<typename Enum, typename = std::enable_if_t<std::is_enum_v<Enum>>> _NODISCARD _CONSTANT_VAR
+etoi(_In_ Enum e) noexcept { return static_cast<std::underlying_type_t<Enum>>(e); }
+
+// function alias template for etoi (EnumToIntegral)
+template<typename Enum> _NODISCARD _CONSTANT_VAR to_integral = etoi<Enum>;
+
+namespace detail {
+
+/// <summary>
+///   Casts <typeparamref name="InputString" /> into an
+///   <typeparamref name="OutputString" />. This won't work with multi-byte
+///   strings, only byte strings. It can be used to convert <c>std::string</c> to
+///   and from <c>std::wstring</c> but that's only if they are byte strings.
+/// </summary>
+/// <typeparam name="OutputString">
+///   The string type specified for conversion.
+/// </typeparam>
+/// <typeparam name="InputString">
+///   The string type to be converted to, as long as an intellisense error
+///   doesn't occur, it should be fine.
+/// </typeparam>
+/// <param name="str">The string to be converted.</param>
+/// <returns>
+///   <typeparamref name="InputString" /> as an
+///   <typeparamref name="OutputString" />.
+/// </returns>
+/// <exception cref="std::ios_base::failure">
+///   If this exception is thrown, this operation as no effect and the
+///   <typeparamref name="InputString" /> is in a valid state.
+/// </exception>
+/// <remarks>
+///   At least <typeparamref name="OutputString" /> needs to be specified and
+///   <typeparamref name="InputString" /> must not be the same as
+///   <typeparamref name="OutputString" />. Use <see cref="ra::Widen" /> or
+///   <see cref="ra::Narrow" /> if you need a Multi-byte to Unicode string
+///   conversion.
+/// </remarks>
+template<
+	typename OutputString,
+	typename InputString,
+	class = std::enable_if_t<(not std::is_same_v<OutputString, InputString>) and
+	(is_char_v<typename OutputString::value_type> and is_char_v<typename InputString::value_type>) and
+	(is_same_size_v<typename OutputString::value_type, typename InputString::value_type>)>
+>
+_NODISCARD OutputString string_cast(const InputString& str) noexcept
+{
+	if (str.empty())
+		return OutputString{}; // an alternative to throwing an exception
+
+	using out_ostringstream = std::basic_ostringstream<typename OutputString::value_type>;
+	out_ostringstream oss;
+	std::ios::sync_with_stdio(false);
+	for (auto& i : str)
```

Comment:
it looks like this function is only used to convert `DataStream` (i.e. byte array) to/from `std::string`. As such, there will never be size converting and using a for loop to copy data seems quite inefficient.

Reply:
Furthermore, `DataStream` is only used by RAWeb as a way to encapsulate the server response, which must support UTF-8. Nothing appears to be converting from std::string back to a `DataStream`. The parameters passed to RAWeb are UTF-8 std::strings.

I'd rather modify RAWeb to return a std::string (assumed to be UTF-8) and eliminate the `DataStream` data type entirely.

Reply:
Yeah it's possible to use <code>std::string</code> for utf-8, there's also a literal <code>u8</code> to force it to be UTF-8 in C++17 such as <code>u8"Hello world!"</code>.

I'm going to see if it will still work with just <code>std::string</code> as I go over this again.